File: src/repl_log.zig
const std = @import("std");
const helpers = @import("helpers.zig");
const log = helpers.log;

pub const ReplOp = enum { put, delete };

pub const ReplEntry = struct {
    id: u64 = 0,
    op: ReplOp,
    path: []const u8,
    completed: bool = false,
    in_flight: bool = false,
};

pub const ReplLog = struct {
    allocator: std.mem.Allocator,
    backing_dir: []const u8,
    mutex: std.Thread.Mutex = .{},
    cond: std.Thread.Condition = .{},
    entries: std.ArrayList(ReplEntry),
    completed_count: usize = 0,
    last_truncate_time: i64 = 0,
    next_id: u64 = 0,
    shutdown: *std.atomic.Value(bool),

    pub fn init(allocator: std.mem.Allocator, backing_dir: []const u8, shutdown: *std.atomic.Value(bool)) !ReplLog {
        var self = ReplLog{
            .allocator = allocator,
            .backing_dir = backing_dir,
            .entries = .empty,
            .shutdown = shutdown,
        };
        self.last_truncate_time = std.time.timestamp();
        self.loadFromDisk() catch |err| {
            log.warn("failed to load replication log: {}", .{err});
        };
        return self;
    }

    pub fn deinitEntries(self: *ReplLog) void {
        for (self.entries.items) |entry| self.allocator.free(entry.path);
        self.entries.deinit(self.allocator);
    }

    fn logPath(self: *ReplLog) ![]const u8 {
        return try std.fs.path.join(self.allocator, &.{ self.backing_dir, ".helmetfs", "repl.log" });
    }

    fn loadFromDisk(self: *ReplLog) !void {
        const path = try self.logPath();
        defer self.allocator.free(path);

        const file = std.fs.openFileAbsolute(path, .{}) catch |err| switch (err) {
            error.FileNotFound => return,
            else => return err,
        };
        defer file.close();

        const contents = file.readToEndAlloc(self.allocator, 16 * 1024 * 1024) catch return error.OutOfMemory;
        defer self.allocator.free(contents);

        var iter = std.mem.splitScalar(u8, contents, '\n');
        while (iter.next()) |line| {
            if (line.len == 0) continue;
            self.parseLine(line) catch continue;
        }

        const pending = self.entries.items.len;
        if (pending > 0) {
            log.info("loaded {} pending replication entries from log", .{pending});
        }
    }

    pub fn parseLine(self: *ReplLog, line: []const u8) !void {
        const first_space = std.mem.indexOfScalar(u8, line, ' ') orelse return error.InvalidFormat;
        const op_str = line[0..first_space];
        const rel_path = line[first_space + 1 ..];

        const op: ReplOp = if (std.mem.eql(u8, op_str, "put"))
            .put
        else if (std.mem.eql(u8, op_str, "delete"))
            .delete
        else
            return error.InvalidOp;

        const id = self.next_id;
        self.next_id += 1;
        try self.entries.append(self.allocator, .{
            .id = id,
            .op = op,
            .path = try self.allocator.dupe(u8, rel_path),
        });
    }

    pub fn enqueue(self: *ReplLog, op: ReplOp, rel_path: []const u8) !void {
        self.mutex.lock();
        defer self.mutex.unlock();

        const path_copy = try self.allocator.dupe(u8, rel_path);
        const id = self.next_id;
        self.next_id += 1;
        self.entries.append(self.allocator, .{ .id = id, .op = op, .path = path_copy }) catch |err| {
            self.allocator.free(path_copy);
            return err;
        };

        self.appendToDisk(op, rel_path) catch |err| {
            log.err("failed to append to replication log: {}", .{err});
        };

        self.cond.signal();
    }

    fn appendToDisk(self: *ReplLog, op: ReplOp, rel_path: []const u8) !void {
        const path = try self.logPath();
        defer self.allocator.free(path);

        const file = try std.fs.createFileAbsolute(path, .{ .truncate = false });
        defer file.close();
        try file.seekFromEnd(0);

        const line = try formatLogEntry(self.allocator, op, rel_path);
        defer self.allocator.free(line);
        try file.writeAll(line);
        try file.sync();
    }

    pub fn dequeueNext(self: *ReplLog) ?struct { id: u64, op: ReplOp, path: []const u8 } {
        self.mutex.lock();
        defer self.mutex.unlock();

        while (!self.shutdown.load(.acquire)) {
            for (self.entries.items, 0..) |*entry, i| {
                if (entry.completed or entry.in_flight) continue;

                if (entry.op == .put) {
                    var dominated = false;
                    for (self.entries.items[i + 1 ..]) |*later| {
                        if (later.op == .put and !later.completed and std.mem.eql(u8, later.path, entry.path)) {
                            dominated = true;
                            break;
                        }
                    }
                    if (dominated) {
                        entry.completed = true;
                        self.completed_count += 1;
                        continue;
                    }
                }

                entry.in_flight = true;
                return .{ .id = entry.id, .op = entry.op, .path = entry.path };
            }
            self.cond.wait(&self.mutex);
        }
        return null;
    }

    pub fn markCompleted(self: *ReplLog, id: u64) void {
        self.mutex.lock();
        defer self.mutex.unlock();

        for (self.entries.items) |*entry| {
            if (entry.id == id) {
                entry.completed = true;
                entry.in_flight = false;
                self.completed_count += 1;
                break;
            }
        }

        self.maybeTruncate();
    }

    pub fn markCompletedByPath(self: *ReplLog, rel_path: []const u8) void {
        self.mutex.lock();
        defer self.mutex.unlock();

        for (self.entries.items) |*entry| {
            if (!entry.completed and std.mem.eql(u8, entry.path, rel_path)) {
                entry.completed = true;
                entry.in_flight = false;
                self.completed_count += 1;
            }
        }
    }

    pub fn pendingCountLocked(self: *ReplLog) u64 {
        var count: u64 = 0;
        for (self.entries.items) |entry| {
            if (!entry.completed) count += 1;
        }
        return count;
    }

    pub fn hasPendingPut(self: *ReplLog, rel_path: []const u8) bool {
        self.mutex.lock();
        defer self.mutex.unlock();
        for (self.entries.items) |entry| {
            if (!entry.completed and entry.op == .put and std.mem.eql(u8, entry.path, rel_path)) {
                return true;
            }
        }
        return false;
    }

    fn maybeTruncate(self: *ReplLog) void {
        const total = self.entries.items.len;
        if (total == 0 or self.completed_count == 0) return;

        const now = std.time.timestamp();
        const should_truncate = (self.completed_count * 2 > total) or
            (now - self.last_truncate_time >= 60);

        if (!should_truncate) return;

        var remaining: std.ArrayList(ReplEntry) = .{};
        for (self.entries.items) |entry| {
            if (!entry.completed) {
                remaining.append(self.allocator, entry) catch {
                    remaining.deinit(self.allocator);
                    log.err("OOM during replication log truncation, skipping", .{});
                    return;
                };
            }
        }

        for (self.entries.items) |entry| {
            if (entry.completed) {
                self.allocator.free(entry.path);
            }
        }

        self.entries.deinit(self.allocator);
        self.entries = remaining;
        self.completed_count = 0;
        self.last_truncate_time = now;

        self.rewriteLogAtomic() catch |err| {
            log.err("failed to truncate replication log: {}", .{err});
        };
    }

    fn rewriteLogAtomic(self: *ReplLog) !void {
        const tmp_path = try std.fs.path.join(self.allocator, &.{ self.backing_dir, ".helmetfs", "repl.log.tmp" });
        defer self.allocator.free(tmp_path);
        const log_path = try self.logPath();
        defer self.allocator.free(log_path);

        const tmp_file = try std.fs.createFileAbsolute(tmp_path, .{});
        defer tmp_file.close();

        for (self.entries.items) |entry| {
            const line = try formatLogEntry(self.allocator, entry.op, entry.path);
            defer self.allocator.free(line);
            try tmp_file.writeAll(line);
        }
        try tmp_file.sync();

        try std.fs.renameAbsolute(tmp_path, log_path);

        if (std.fs.path.dirname(log_path)) |dir_path| {
            helpers.fsyncDir(dir_path);
        }
    }
};

pub fn formatLogEntry(allocator: std.mem.Allocator, op: ReplOp, rel_path: []const u8) ![]const u8 {
    return try std.fmt.allocPrint(allocator, "{s} {s}\n", .{ @tagName(op), rel_path });
}

test "formatLogEntry/parseLine round-trip" {
    const testing = std.testing;
    const allocator = testing.allocator;

    // Create a minimal ReplLog for parseLine testing
    var shutdown = std.atomic.Value(bool).init(false);
    var tmp_dir_buf: ["/tmp/helmetfs-rl-test-XXXXXX".len:0]u8 = "/tmp/helmetfs-rl-test-XXXXXX".*;
    const result = helpers.c.mkdtemp(&tmp_dir_buf);
    if (result == null) return error.TmpDirFailed;
    const tmp_dir = try allocator.dupe(u8, std.mem.span(result));
    defer allocator.free(tmp_dir);
    defer std.fs.deleteTreeAbsolute(tmp_dir) catch {};

    const helmetfs_dir = try std.fs.path.join(allocator, &.{ tmp_dir, ".helmetfs" });
    defer allocator.free(helmetfs_dir);
    try std.fs.makeDirAbsolute(helmetfs_dir);

    var rl = try ReplLog.init(allocator, tmp_dir, &shutdown);
    defer rl.deinitEntries();

    const cases = .{
        .{ .op = ReplOp.put, .path = "hello/world.txt" },
        .{ .op = ReplOp.delete, .path = "gone.txt" },
    };
    inline for (cases) |tc| {
        const line = try formatLogEntry(allocator, tc.op, tc.path);
        defer allocator.free(line);
        const trimmed = std.mem.trimRight(u8, line, "\n");
        const before = rl.entries.items.len;
        try rl.parseLine(trimmed);
        try testing.expectEqual(before + 1, rl.entries.items.len);
        const entry = rl.entries.getLast();
        try testing.expectEqual(tc.op, entry.op);
        try testing.expectEqualStrings(tc.path, entry.path);
    }
}

File: src/repl_log.rs
//! Replication log: in-memory queue with disk persistence.
//!
//! Each entry is either `put <path>` or `delete <path>`.
//! Put entries support coalescing: when a new put is enqueued for a path that
//! already has a pending (not-yet-started) put, the earlier entry is marked
//! completed so the replication worker skips it.  Deletes are never coalesced.
//!
//! On-disk format (`<backing>/.helmetfs/repl.log`):
//!   put <path>\n
//!   delete <path>\n

use std::collections::VecDeque;
use std::fs;
use std::io::{self, BufRead, Write};
use std::path::{Path, PathBuf};
use std::sync::{Condvar, Mutex};

// ---------------------------------------------------------------------------
// Entry types
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum ReplOp {
    Put,
    Delete,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct ReplEntry {
    pub op: ReplOp,
    pub path: String,
    /// Stable identifier that never changes once assigned.
    pub id: u64,
    /// Set to true when a worker has claimed this entry via wait_next/try_next.
    pub in_progress: bool,
    /// Set to true when the entry has been processed (or coalesced away).
    pub completed: bool,
}

// ---------------------------------------------------------------------------
// ReplQueue — pure replication queue logic (no I/O, no synchronization)
// ---------------------------------------------------------------------------

/// Pure replication queue logic shared between the real implementation and
/// the stateright model.  Contains no I/O, locking, or persistence.
#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct ReplQueue {
    pub entries: VecDeque<ReplEntry>,
    /// Monotonically increasing counter for assigning stable entry IDs.
    pub next_id: u64,
}

impl Default for ReplQueue {
    fn default() -> Self {
        Self::new()
    }
}

impl ReplQueue {
    pub fn new() -> Self {
        Self {
            entries: VecDeque::new(),
            next_id: 0,
        }
    }

    /// Enqueue a `put` entry. Coalesces with any existing uncompleted entry for
    /// the same path (marks earlier ones completed).
    ///
    /// This coalesces both earlier puts (same-op) AND earlier deletes (cross-op).
    /// A put copies the current backing content, so it fully subsumes an
    /// earlier delete—the replica will end up with the latest file regardless.
    /// Without cross-op coalescing, multi-worker out-of-order processing can
    /// cause an earlier delete to execute after this put, leaving the replica
    /// empty.
    pub fn enqueue_put(&mut self, path: &str) {
        for entry in self.entries.iter_mut() {
            if !entry.completed && !entry.in_progress && entry.path == path {
                entry.completed = true;
            }
        }
        let id = self.next_id;
        self.next_id += 1;
        self.entries.push_back(ReplEntry {
            op: ReplOp::Put,
            path: path.to_string(),
            id,
            in_progress: false,
            completed: false,
        });
    }

    /// Enqueue a `delete` entry. Cross-op coalescing: mark earlier pending
    /// (not in_progress) puts for the same path as completed—there is no point
    /// copying a file we are about to remove.
    pub fn enqueue_delete(&mut self, path: &str) {
        for entry in self.entries.iter_mut() {
            if !entry.completed && !entry.in_progress && entry.op == ReplOp::Put && entry.path == path {
                entry.completed = true;
            }
        }
        let id = self.next_id;
        self.next_id += 1;
        self.entries.push_back(ReplEntry {
            op: ReplOp::Delete,
            path: path.to_string(),
            id,
            in_progress: false,
            completed: false,
        });
    }

    /// Check if there is a pending (uncompleted) `put` for the given path.
    /// Entries that are in_progress are also considered pending (being processed).
    pub fn has_pending_put(&self, path: &str) -> bool {
        self.entries
            .iter()
            .any(|e| !e.completed && e.op == ReplOp::Put && e.path == path)
    }

    /// Claim next available entry (not completed, not in_progress).
    /// Marks it as in_progress and returns its stable id and a clone.
    pub fn claim_next(&mut self) -> Option<(u64, ReplEntry)> {
        for entry in self.entries.iter_mut() {
            if !entry.completed && !entry.in_progress {
                entry.in_progress = true;
                return Some((entry.id, entry.clone()));
            }
        }
        None
    }

    /// Claim next available entry, returning only its stable id.
    pub fn claim_next_id(&mut self) -> Option<u64> {
        for entry in self.entries.iter_mut() {
            if !entry.completed && !entry.in_progress {
                entry.in_progress = true;
                return Some(entry.id);
            }
        }
        None
    }

    /// Find entry by stable id.
    pub fn find_entry(&self, id: u64) -> Option<&ReplEntry> {
        self.entries.iter().find(|e| e.id == id)
    }

    /// Mark entry with the given stable `id` as completed and garbage-collect
    /// leading completed entries.
    pub fn mark_completed(&mut self, id: u64) {
        for entry in self.entries.iter_mut() {
            if entry.id == id {
                entry.completed = true;
                entry.in_progress = false;
                break;
            }
        }
        // GC: remove completed entries from the front
        while self
            .entries
            .front()
            .map_or(false, |e| e.completed)
        {
            self.entries.pop_front();
        }
    }

    /// Number of pending (uncompleted) entries.
    pub fn pending_count(&self) -> usize {
        self.entries.iter().filter(|e| !e.completed).count()
    }

    /// IDs of in-progress entries (used for action generation in model).
    pub fn in_progress_ids(&self) -> Vec<u64> {
        self.entries
            .iter()
            .filter(|e| e.in_progress && !e.completed)
            .map(|e| e.id)
            .collect()
    }

    /// Whether there are claimable entries (not completed, not in_progress).
    pub fn has_claimable(&self) -> bool {
        self.entries.iter().any(|e| !e.completed && !e.in_progress)
    }

    /// Reset all in_progress flags (used for crash recovery).
    pub fn reset_in_progress(&mut self) {
        for entry in self.entries.iter_mut() {
            entry.in_progress = false;
        }
    }
}

// ---------------------------------------------------------------------------
// ReplLog — thread-safe wrapper with persistence
// ---------------------------------------------------------------------------

pub struct ReplLog {
    inner: Mutex<ReplQueue>,
    cond: Condvar,
    log_path: PathBuf,
}

impl ReplLog {
    /// Create a new replication log. If `log_path` exists on disk, its entries
    /// are loaded (providing crash-recovery).
    pub fn new(backing_dir: &Path) -> io::Result<Self> {
        let helmetfs_dir = backing_dir.join(".helmetfs");
        fs::create_dir_all(&helmetfs_dir)?;
        let log_path = helmetfs_dir.join("repl.log");

        let mut queue = ReplQueue::new();
        if log_path.exists() {
            let file = fs::File::open(&log_path)?;
            for line in io::BufReader::new(file).lines() {
                let line = line?;
                let line = line.trim();
                if line.is_empty() {
                    continue;
                }
                if let Some(path) = line.strip_prefix("put ") {
                    queue.entries.push_back(ReplEntry {
                        op: ReplOp::Put,
                        path: path.to_string(),
                        id: queue.next_id,
                        in_progress: false,
                        completed: false,
                    });
                    queue.next_id += 1;
                } else if let Some(path) = line.strip_prefix("delete ") {
                    queue.entries.push_back(ReplEntry {
                        op: ReplOp::Delete,
                        path: path.to_string(),
                        id: queue.next_id,
                        in_progress: false,
                        completed: false,
                    });
                    queue.next_id += 1;
                }
            }
            if !queue.entries.is_empty() {
                log::info!("Loaded {} entries from replication log", queue.entries.len());
            }
        }

        Ok(Self {
            inner: Mutex::new(queue),
            cond: Condvar::new(),
            log_path,
        })
    }

    /// Enqueue a `put` entry. Coalesces with any existing uncompleted put for
    /// the same path (marks the earlier one completed).
    pub fn enqueue_put(&self, path: &str) {
        let mut queue = self.inner.lock().unwrap();
        queue.enqueue_put(path);
        self.persist_locked(&queue);
        self.cond.notify_one();
    }

    /// Enqueue a `delete` entry. Cross-op coalescing: mark earlier pending
    /// (not in_progress) puts for the same path as completed.
    pub fn enqueue_delete(&self, path: &str) {
        let mut queue = self.inner.lock().unwrap();
        queue.enqueue_delete(path);
        self.persist_locked(&queue);
        self.cond.notify_one();
    }

    /// Block until there is an uncompleted, not-in-progress entry, then mark it
    /// as in_progress and return a clone of it (and its stable id). The caller
    /// must call `mark_completed` after processing.
    /// Returns `None` if woken up with no entries (e.g. shutdown signal).
    pub fn wait_next(&self) -> Option<(u64, ReplEntry)> {
        let mut queue = self.inner.lock().unwrap();
        loop {
            if let Some(result) = queue.claim_next() {
                return Some(result);
            }
            // Use wait_timeout to avoid permanent blocking
            let (guard, _timeout) = self
                .cond
                .wait_timeout(queue, std::time::Duration::from_millis(200))
                .unwrap();
            queue = guard;

            // After wakeup, check again — if still nothing, return None
            // so the caller can check for shutdown
            if !queue.has_claimable() {
                return None;
            }
        }
    }

    /// Try to get next uncompleted, not-in-progress entry without blocking.
    /// Marks it as in_progress. Returns None if there are no available entries.
    pub fn try_next(&self) -> Option<(u64, ReplEntry)> {
        let mut queue = self.inner.lock().unwrap();
        queue.claim_next()
    }

    /// Mark entry with the given stable `id` as completed and garbage-collect
    /// leading completed entries.
    pub fn mark_completed(&self, id: u64) {
        let mut queue = self.inner.lock().unwrap();
        queue.mark_completed(id);
        self.persist_locked(&queue);
    }

    /// Check if there is a pending (uncompleted) `put` for the given path.
    /// Entries that are in_progress are also considered pending (being processed).
    pub fn has_pending_put(&self, path: &str) -> bool {
        let queue = self.inner.lock().unwrap();
        queue.has_pending_put(path)
    }

    /// Wake up all waiting workers (used during shutdown).
    pub fn notify_all(&self) {
        self.cond.notify_all();
    }

    // -----------------------------------------------------------------------
    // Persistence
    // -----------------------------------------------------------------------

    fn persist_locked(&self, queue: &ReplQueue) {
        if let Err(e) = self.persist_impl(queue) {
            log::error!("Failed to persist replication log: {}", e);
        }
    }

    fn persist_impl(&self, queue: &ReplQueue) -> io::Result<()> {
        let tmp = self.log_path.with_extension("log.tmp");
        {
            let mut f = fs::File::create(&tmp)?;
            for entry in &queue.entries {
                if entry.completed {
                    continue;
                }
                match entry.op {
                    ReplOp::Put => writeln!(f, "put {}", entry.path)?,
                    ReplOp::Delete => writeln!(f, "delete {}", entry.path)?,
                }
            }
            f.sync_all()?;
        }
        fs::rename(&tmp, &self.log_path)?;
        Ok(())
    }
}
